versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2022 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.adapter.spi;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Consumer;
+import org.agrona.collections.Hashing;
+import org.agrona.collections.Object2IntHashMap;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.Key;
+import org.projectnessie.versioned.persist.adapter.CommitLogEntry;
+import org.projectnessie.versioned.persist.adapter.KeyListEntity;
+import org.projectnessie.versioned.persist.adapter.KeyListEntry;
+
+final class FetchValuesUsingOpenAddressing {
+
+  final int[] keyListSegmentOffsets;
+  final int keyListCount;
+  final KeyListEntry[][] keyListsArray;
+  final Object2IntHashMap<Hash> entityIdToSegment;
+  final int hashMask;
+  final List<Hash> keyListIds;
+
+  FetchValuesUsingOpenAddressing(CommitLogEntry entry) {
+    List<Integer> keyListSegmentOffsetsList = entry.getKeyListEntityOffsets();
+    keyListSegmentOffsets =
+        keyListSegmentOffsetsList != null
+            ? keyListSegmentOffsetsList.stream().mapToInt(Integer::intValue).toArray()
+            : new int[0];
+    keyListCount = 1 + keyListSegmentOffsets.length;
+
+    keyListsArray = new KeyListEntry[keyListCount][];
+    keyListsArray[0] = entry.getKeyList().getKeys().toArray(new KeyListEntry[0]);
+
+    entityIdToSegment = new Object2IntHashMap<>(keyListCount, Hashing.DEFAULT_LOAD_FACTOR, -1);
+
+    hashMask = entry.getKeyListBucketCount() - 1;
+
+    keyListIds = entry.getKeyListsIds();
+  }
+
+  /** Calculates the open-addressing bucket for a key. */
+  int bucketForKey(Key key) {
+    return key.hashCode() & hashMask;
```

Comment:
nit: Is `%` much slower? Using `%` would remove the dependency on the hash table size being a power of 2 on reads :thinking: